### PR TITLE
Fixes ASM change for ResourcesCompat

### LINKED ISSUE #1617: Paparazzi 1.3.5-SNAPSHOT breaks Bouncycastle class loading
**Description**
Paparazzi 1.3.5-SNAPSHOT contains a [fix](https://github.com/cashapp/paparazzi/commit/c9a5b2ef5f9b74689400ed36a9a6a49032e43dfe) for font loading. It hotpatches a class to achieve this.

Unfortunately, there's an issue: The following line apparently causes it to instrument _all_ classes, which breaks the signature of Bouncycastle classes:

```kotlin
override fun isInstrumentable(classData: ClassData): Boolean = true
```

We are seeing test failures in unrelated Robolectric tests with the following error:

```
Could not initialize class org.bouncycastle.jcajce.provider.asymmetric.compositesignatures.CompositeSignaturesConstants
java.lang.NoClassDefFoundError: Could not initialize class org.bouncycastle.jcajce.provider.asymmetric.compositesignatures.CompositeSignaturesConstants
...
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.SecurityException: SHA-256 digest error for org/bouncycastle/jcajce/provider/asymmetric/compositesignatures/CompositeSignaturesConstants$CompositeName.class [in thread "SDK 34 Main Thread"]
```

**Suggested Fix**

I've applied the following change in a bugfix fork of Paparazzi and it seems to work:

```kotlin
override fun isInstrumentable(classData: ClassData): Boolean =
  classData.className == ResourcesCompatTransform.RESOURCES_COMPAT_CLASS_NAME.replace('/', '.')
```

Even though there's a check in `visitMethod` already for this class, it seems like even instrumenting Bouncycastle without making any actual changes to the bytecode breaks its signature.

**Steps to Reproduce**
- Create a Robolectric test in a module that also contains Paparazzi tests
- Make sure the Paparazzi tests run before the Robolectric tests
- Observe errors like `java.lang.SecurityException at ManifestEntryVerifier.java:256` with the root cause explained above

**Expected behavior**
Paparazzi should only instrument the `ResourcesCompat` class.

**Additional information:**
- Paparazzi Version: 1.3.5-SNAPSHOT from Oct 2, 2024
- OS: Linux
- Compile SDK: 34
- Gradle Version: 8.7
- Android Gradle Plugin Version: 8.6.0

### PR DESCRIPTION
**Pull request by geoff-powell**
Fixes #1617

### CODE REVIEW

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/instrumentation/ResourcesCompatVisitorFactory.kt [hunk 1]
```diff
@@ -73,10 +62,13 @@ internal abstract class ResourcesCompatVisitorFactory : AsmClassVisitorFactory<I
...
 
     private companion object {
-      const val RESOURCES_COMPAT_CLASS_NAME = "androidx/core/content/res/ResourcesCompat"
       const val LOAD_FONT_METHOD_NAME = "loadFont"
       const val LOAD_FONT_METHOD_DESCRIPTOR =
         "(Landroid/content/Context;Landroid/content/res/Resources;Landroid/util/TypedValue;IILandroidx/core/content/res/ResourcesCompat\$FontCallback;Landroid/os/Handler;ZZ)Landroid/graphics/Typeface;"
     }
   }
+
+  internal companion object {
```
Comment (geoff-powell) [inline paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/instrumentation/ResourcesCompatVisitorFactory.kt:71]: nit: private

File: paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt [hunk 2]
```diff
@@ -1505,6 +1505,15 @@ class PaparazziPluginTest {
       .runFixture(fixtureRoot) { build() }
   }
 
+  @Test
+  fun robolectric() {
+    val fixtureRoot = File("src/test/projects/robolectric")
+
+    gradleRunner
+      .withArguments("verifyPaparazziDebug", "--stacktrace")
```
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt:1513]:
```suggestion
      .withArguments("testPaparazziDebug", "--stacktrace")
```

no need to verify, this test only wants to ensure Paparazzi's ASM + Robolectric can co-exist.

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/instrumentation/ResourcesCompatVisitorFactory.kt [hunk 3]
```diff
@@ -27,19 +29,6 @@ internal abstract class ResourcesCompatVisitorFactory : AsmClassVisitorFactory<I
...
-      version: Int,
-      access: Int,
-      name: String,
-      signature: String?,
-      superName: String?,
-      interfaces: Array<String>?
-    ) {
-      isResourcesCompatClass = name == RESOURCES_COMPAT_CLASS_NAME
-      super.visit(version, access, name, signature, superName, interfaces)
-    }
```
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/instrumentation/ResourcesCompatVisitorFactory.kt:32]:
```suggestion
```

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/instrumentation/ResourcesCompatVisitorFactory.kt (hunk 1 above)
Comment (jrodbx) [reply paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/instrumentation/ResourcesCompatVisitorFactory.kt:71]: nit: keep the 3 consts together, but move the `private companion` from the inner class to the outer class level

File: paparazzi-gradle-plugin/src/test/projects/robolectric/src/test/snapshots/images/app.cash.paparazzi.plugin.test_PaparazziTest_test.png
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/projects/robolectric/src/test/snapshots/images/app.cash.paparazzi.plugin.test_PaparazziTest_test.png:1]: delete

File: paparazzi-gradle-plugin/src/test/projects/robolectric/src/test/java/app/cash/paparazzi/plugin/test/RobolectricTest.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class RobolectricTest {
+  @Test
+  fun test() {
+    Robolectric.buildActivity(MainActivity::class.java).use { controller ->
+      controller.setup()
+    }
+  }
```
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/projects/robolectric/src/test/java/app/cash/paparazzi/plugin/test/RobolectricTest.kt:30]:
```suggestion
  fun test() = Unit
```

reducing to a minimal test highlights the core issue here.  it's merely the presence of both libraries on the classpath interfering with this runner's bootstrap.

File: paparazzi-gradle-plugin/src/test/projects/robolectric/src/test/java/app/cash/paparazzi/plugin/test/PaparazziTest.kt
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/projects/robolectric/src/test/java/app/cash/paparazzi/plugin/test/PaparazziTest.kt:1]: this file can be deleted, as the underlying issue occurs regardless of the instantiation of the rule or the call to snapshot.

File: paparazzi-gradle-plugin/src/test/projects/robolectric/src/main/java/app/cash/paparazzi/plugin/test/MainActivity.kt
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/projects/robolectric/src/main/java/app/cash/paparazzi/plugin/test/MainActivity.kt:1]: delete

**Review by jrodbx — APPROVED** [review]

**Pull request merged**